File: RoundRobin.java
import java.util.Collections;
import java.util.Comparator;
import java.util.Arrays;

class RoundRobin {
    private Prisoner[] prisoners;
    private int bouts;

    public RoundRobin(Prisoner[] prisoners, int bouts) {
        //Constructor
        this.prisoners = prisoners;
        this.bouts = bouts;
    }

    public void run() {
        this.run(false);
    }

    public void run(boolean verbose) {
        //Run 'bouts' bouts for every pairing of prisoner, including themselves
        System.out.println("Running a round robin tournament with " + bouts + " bouts between " + prisoners.length + " prisoners");
        for (int i = 0; i < prisoners.length; i ++) {
            //Updated to make prisoners play themselves
            for (int j = i; j < prisoners.length; j ++) {
                Match match = new Match(prisoners[i], prisoners[j], bouts);
                match.play(verbose);
            }
        }
    }

    public void sort() {
        //Sort the prisoners in order of score, using the inbuilt sort method
        Collections.sort(Arrays.asList(prisoners), new Comparator<Prisoner>() {
            public int compare(Prisoner p1, Prisoner p2) {
                int score1 = p1.getScore();
                int score2 = p2.getScore();
                if (score1 > score2) {
                    return -1;
                }
                else if (score1 == score2) {
                    return 0;
                }
                else {
                    return 1;
                }
            }
        });
    }

    public void displayStats() {
        //Print out the league table
        String[] data = new String[prisoners.length];
        int longest = 0;
        for(int i = 0; i < prisoners.length; i ++) {
            String prisonerData = prisoners[i].toString();
            int length = prisonerData.length();
            if (length > longest) {
                longest = length;
            }
            data[i] = prisonerData;
        }
        Table table = new Table(data, "Round Robin Results", longest);
        System.out.println(table);
    }
}

File: RoundRobin.kt
package de.freyama.prisoners_dilemma
import de.freyama.prisoners_dilemma.prisoner.*

/**
 * A RoundRobin instance represents a tournament played in RoundRobin style between a group of Prisoners.
 * Each pairing of Prisoners will play a Match with the specified number of bouts, then a league table will be drawn.
 */
class RoundRobin(private val prisoners: Array<Prisoner>, private val bouts: Int) {

    /**
     * Run the tournament using the setup specified.
     */
    fun run(verbose: Boolean = false) {
        // Print out a header for the tournament.
        println("Running a round robin tournament, with $bouts bouts per match, and ${this.prisoners.size} prisoners.")
        for (i in 0..(this.prisoners.size - 1)) {
            for (j in i..(this.prisoners.size - 1)) {
                val match = Match(this.prisoners[i], this.prisoners[j], bouts)
                match.play(verbose)
            }
        }

        // After the tournament is over, print out a table from highest to lowest points.
        // Sort the prisoners Array
        this.prisoners.sortBy({prisoner -> -prisoner.score})

        // Draw a nice table
        this.renderTable()
    }

    private fun renderTable() {
        // Firstly, generate the column sizes.
        val column1Size = this.prisoners.map({prisoner -> prisoner.strategy.length}).max()!!
        var maxScore = this.prisoners.map({prisoner -> prisoner.score}).max()!!
        var column2Size = 0
        while (maxScore > 0) {
            maxScore = maxScore / 10
            column2Size++
        }

        // Print the table
        val row = "│ %-${column1Size}s │ %${column2Size}s │"
        println("┌${"─".repeat(column1Size + 2)}┬${"─".repeat(column2Size + 2)}┐")
        for (prisoner in this.prisoners) {
            println(row.format(prisoner.strategy, prisoner.score))
        }
        println("└${"─".repeat(column1Size + 2)}┴${"─".repeat(column2Size + 2)}┘")
    }
}
